### PR TITLE
Simple get and set on heapValues.

### PR DESCRIPTION
**Pull request by arjunguha**

### CODE REVIEW

**Comment by arjunguha** [comment]
The reason the test fails `KUpdateKey` doesn't preserve the old value.

**Comment by arjunguha** [comment]
@rolivia would you look at the failing test here and give an opinion on what it should produce?

**Comment by rolivia** [comment]
after the conditional update (assuming the guard is true), obj.x should be something like:  <Untrusted ? {y : 200} : {y : 100}>  and the test should produce <Untrusted ? 200 : 100>

**Comment by arjunguha** [comment]
This could quickly turn into `FacetProp` being a new kind of `ObjPropertyV`. I don't think that's the way to go. I think a better approach is for the `set` method of heap values to produce new heap values. Finally, the `KUpdateKey` case could actually update the heap. This would work well for `VObj`, but arrays and native objects are more problematic.

File: js/src/main/scala/HeapValues.scala [hunk 1]
```diff
@@ -160,12 +176,15 @@ trait HeapValues extends slogging.LazyLogging {
...
-      VConfig(jsToValue(obj.selectDynamic(anyToStr(field))), k)
+    def get(addr: VHeapRef, field: Value): Expr = {
+      ValueExpr(jsToValue(obj.selectDynamic(anyToStr(field))))
     }
 
     def set(field: Value, value: Value): Unit = {
-      obj.updateDynamic(anyToStr(field))(valueToJs(value))
+      val key = anyToStr(field)
+      val oldValue = jsToValue(obj.selectDynamic(key))
+      val newValue = newFacet(store.pc, value, oldValue)
```
Comment (rachitnigam) [inline js/src/main/scala/HeapValues.scala]: `remoteCallback` might break this. But then, `remoteCallback` already breaks so many things.

File: js/src/main/scala/HeapValues.scala [hunk 2]
```diff
@@ -126,26 +133,35 @@ trait HeapValues extends slogging.LazyLogging {
...
     }
 
     def set(field: Value, value: Value): Unit = field match {
       case VConst(CNum(x)) if (x.toInt == x && x >= 0) => {
-        items.update(x.toInt, value)
+        val oldValue = try {
+          items.apply(x.toInt)
+        } catch {
+          case (_: scala.scalajs.runtime.UndefinedBehaviorError) =>
+            VBot
```
Comment (rachitnigam) [inline js/src/main/scala/HeapValues.scala]: scalajs returns undefined (actually, it errors out) for an index that has not yet been assigned. Does returning `VBot` look right? Or should it be `CUndefined`.

**Pull request merged**